fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/impl/FHIRPersistenceJDBCImpl.java

```diff
@@ -1800,39 +1795,67 @@ private boolean isWholeSystem(SearchParameter sp) {
         }
+
         Map<String,Set<String>> compartmentRefParams = CompartmentUtil.getCompartmentParamsForResourceType(resourceType);
-        Map<String, Set<CompartmentReference>> compartmentMap = SearchUtil.extractCompartmentParameterValues(fhirResource, compartmentRefParams);
+        Map<String, List<ExtractedParameterValue>> collectedEPVByCode = allParameters.stream()
+                .collect(Collectors.groupingBy(epv -> epv.getName()));
 
-        for (Map.Entry<String, Set<CompartmentReference>> entry: compartmentMap.entrySet()) {
-            final String compartmentName = entry.getKey();
-            final String parameterName = CompartmentUtil.makeCompartmentParamName(compartmentName);
+        for (Entry<String, Set<String>> entry : compartmentRefParams.entrySet()) {
+            String param = entry.getKey();
+            Set<String> compartments = entry.getValue();
 
-            // Create a reference parameter value for each CompartmentReference extracted from the resource
-            for (CompartmentReference compartmentRef: entry.getValue()) {
-                ReferenceParmVal pv = new ReferenceParmVal();
-                pv.setName(parameterName);
-                pv.setResourceType(resourceType);
+            List<ExtractedParameterValue> collectedEPV = collectedEPVByCode.get(param);
+            // If the parameter has no corresponding extracted values, log and continue
+            if (collectedEPV == null) {
+                if (log.isLoggable(Level.FINE)) {
+                    log.warning("Compartment inclusion param " + param + " has no value");
+                }
+                continue;
+            }
 
-                // ReferenceType doesn't really matter here, but LITERAL_RELATIVE is appropriate
-                ReferenceValue rv = new ReferenceValue(compartmentName, compartmentRef.getReferenceResourceValue(), ReferenceType.LITERAL_RELATIVE, null);
-                pv.setRefValue(rv);
+            for (ExtractedParameterValue epv : collectedEPV) {
+                if (!(epv instanceof ReferenceParmVal)) {
+                    log.warning("Skipping compartment inclusion param " + param + "; expected ReferenceParmVal but found " +
+                            epv.getClass().getSimpleName());
+                    continue;
+                }
 
-                if (log.isLoggable(Level.FINE)) {
-                    log.fine("Adding compartment reference parameter: [" + resourceType + "] "+ parameterName + " = " + rv.getTargetResourceType() + "/" + rv.getValue());
+                ReferenceValue rv = ((ReferenceParmVal) epv).getRefValue();
+                if (rv != null && rv.getType() == ReferenceType.LITERAL_RELATIVE
+                        && compartments.contains(rv.getTargetResourceType())) {
+                    String internalCompartmentParamName = CompartmentUtil.makeCompartmentParamName(rv.getTargetResourceType());
+
+                    if (epv.isForStoring()) {
```

Comment:
The code looks good.  I just don't quite follow the "isForStoring" meaning.  Store where, etc?  I think I just need some extra education on the side for background.

Reply:
for storing in the search parameter tables in our db

Reply:
basically this was just the cleanest way I could think of the get some parameters extracted from our existing SearchUtil.extractParameters method without actually paying the cost of storing them in the db